File: VSCode/Codigo01_MODELONAOLINEAR.py
import tclab
import numpy as np
import time
import matplotlib.pyplot as plt

# Conexão com o TCLab
a = tclab.TCLab()

# Tempo de execução em minutos
run_time = 20

# Número de ciclos (1 ciclo por segundo)
loops = int(60.0 * run_time)
tm = np.zeros(loops)            # Vetor para armazenar o tempo
T1 = np.zeros(loops)            # Temperatura real
T1_ModeloNL = np.zeros(loops)   # Temperatura do modelo não linear
Q1 = np.ones(loops) * 50        # Potência fixa em 50%

# Aplicar potência inicial
a.Q1(50)

# Criar gráfico
plt.figure(figsize=(12, 8))
plt.ion()
plt.show()

# Loop principal
start_time = time.time()
prev_time = start_time

try:
    for i in range(loops):
        # Tempo de espera
        sleep_time = max(0.01, 1.0 - (time.time() - prev_time))
        time.sleep(sleep_time)

        # Atualizar tempo
        t = time.time()
        prev_time = t
        tm[i] = t - start_time

        # Ler temperatura real do TCLab
        T1[i] = a.T1

        # Informações para o modelo
        Taquecedor = T1_ModeloNL[i-1] + 273.15 if i > 0 else a.T1 + 273.15  # Temperatura simulada em Kelvin
        Tambiente = 301.15  # Temperatura ambiente (28°C em Kelvin)
        Alpha = 0.01
        Cp = 500
        A = 0.0012
        m = 0.004
        U = 10
        Emissividade = 0.9
        Boltzmann = 5.67e-8
        Q = 50
        dt = 1.0  # intervalo de tempo em segundos

        # Modelo não linear: cálculo de dT/dt
        dTdt = ((Alpha / (m * Cp)) * Q) + ((U * A) / (m * Cp)) * (Tambiente - Taquecedor) + ((Emissividade * Boltzmann * A) / (m * Cp)) * (Tambiente**4 - Taquecedor**4)

        if i == 0:
            T1_ModeloNL[i] = a.T1  # Condição inicial: T real no instante 0
        else:
            T1_ModeloNL[i] = T1_ModeloNL[i-1] + dTdt * dt  # Temperatura estimada

        # Atualizar gráfico
        plt.clf()
        ax1 = plt.subplot(2, 1, 1)
        ax1.grid()
        plt.plot(tm[:i+1], T1[:i+1], 'r-', label='Temperatura Real', linewidth=2)
        plt.plot(tm[:i+1], T1_ModeloNL[:i+1], 'b--', label='Modelo Não Linear', linewidth=2)
        plt.ylabel('Temperatura (°C)', fontsize=14)
        plt.legend(loc='best', fontsize=12)

        ax2 = plt.subplot(2, 1, 2)
        ax2.grid()
        plt.plot(tm[:i+1], Q1[:i+1], 'r-', label='Potência (%)', linewidth=2)
        plt.ylabel('Potência (%)', fontsize=14)
        plt.xlabel('Tempo (s)', fontsize=14)
        plt.legend(loc='best', fontsize=12)

        plt.tight_layout()
        plt.draw()
        plt.pause(0.05)

    # --- Cálculo do Erro e Salvamento ---

    # Erro Médio Absoluto
    erro_medio = np.mean(np.abs(T1 - T1_ModeloNL))
    print(f'\nErro Médio Absoluto entre Temperatura Real e Modelo: {erro_medio:.2f} °C')

    # Salvar gráfico em PNG
    plt.savefig('Grafico_TCLab_ModeloNL.png', dpi=300, bbox_inches='tight')

    # Salvar dados em .txt
    np.savetxt('Dados_simulacao_01.txt', np.column_stack((tm, T1, T1_ModeloNL, Q1)),
               header='Tempo (s)\tTemperatura Real (°C)\tTemperatura Modelo (°C)\tPotência (%)', fmt='%.2f', delimiter='\t')

# Capturar interrupção manual (Ctrl+C)
except KeyboardInterrupt:
    print('Interrompido pelo usuário.')

# Capturar outros erros
except Exception as e:
    print(f'Erro: {e}')
    raise

# Finalizar e garantir que o aquecedor desligue
finally:
    a.Q1(0)    # Desliga o aquecedor
    a.LED(0)   # Apaga LED
    a.close()
    print('Simulação concluída.')

File: MATLAB/Codigo01_Modelo_nao_linear.m
clear; clc;

%% Inicialização do hardware
run('tclab.m'); % Arquivo com funções T1C(), h1(), led()

% Garantir desligamento ao final, mesmo com Ctrl+C ou erro
cleanupObj = onCleanup(@() desligarTCLab());

%% Parâmetros de Simulação
run_time = 20;                 % Tempo total da simulação (min)
loops = 60 * run_time;         % Número de amostras (1 Hz)
dt = 1.0;                      % Intervalo de tempo (s)

% Inicialização de variáveis
tm = zeros(loops,1);           % Vetor de tempo (s)
T1 = zeros(loops,1);           % Temperatura real (°C)
T1_ModeloNL = zeros(loops,1);  % Temperatura estimada (modelo NL)
Q1 = ones(loops,1) * 50;       % Potência aplicada (%)

% Aplicar potência inicial
h1(Q1(1));

%% Parâmetros físicos do modelo
Tambiente = 301.15;     % Temperatura ambiente (K)
Alpha = 0.01;           % Eficiência do aquecedor
Cp = 500;               % Calor específico (J/kg.K)
A = 0.0012;             % Área de troca térmica (m²)
m = 0.004;              % Massa (kg)
U = 10;                 % Coef. convecção (W/m²K)
Emissividade = 0.9;     % Emissividade
Boltzmann = 5.67e-8;    % Constante de Stefan-Boltzmann

%% Criação do gráfico
figure('Position', [100 100 800 600]);

%% Loop principal de aquisição e simulação
start_time = tic;
prev_time = toc(start_time);

for i = 1:loops
    % Espera até completar 1 segundo
    sleep_time = max(0.01, 1.0 - (toc(start_time) - prev_time));
    pause(sleep_time);

    % Atualizar tempo
    t = toc(start_time);
    prev_time = t;
    tm(i) = t;

    % Leitura da temperatura real
    T1(i) = T1C();

    % Atualização do modelo não linear
    if i == 1
        T1_ModeloNL(i) = T1(i);
    else
        Taquecedor = T1_ModeloNL(i-1) + 273.15; % Converter para Kelvin
        Q = Q1(i);
        
        % Equação diferencial do modelo
        dTdt = ((Alpha / (m * Cp)) * Q) ...
             + ((U * A) / (m * Cp)) * (Tambiente - Taquecedor) ...
             + ((Emissividade * Boltzmann * A) / (m * Cp)) * (Tambiente^4 - Taquecedor^4);
        
        % Integração de Euler
        T1_ModeloNL(i) = T1_ModeloNL(i-1) + dTdt * dt;
    end

    % Atualização gráfica
    clf;
    subplot(2,1,1);
    plot(tm(1:i), T1(1:i), 'r-', 'LineWidth', 2); hold on;
    plot(tm(1:i), T1_ModeloNL(1:i), 'b--', 'LineWidth', 2);
    ylabel('Temperatura (°C)', 'FontSize', 14);
    legend('Temperatura Real', 'Modelo Não Linear', 'FontSize', 12);
    grid on;

    subplot(2,1,2);
    plot(tm(1:i), Q1(1:i), 'r-', 'LineWidth', 2);
    ylabel('Potência (%)', 'FontSize', 14);
    xlabel('Tempo (s)', 'FontSize', 14);
    legend('Potência (%)', 'FontSize', 12);
    grid on;

    drawnow;
end

%% Pós-processamento
erro_medio = mean(abs(T1 - T1_ModeloNL));
fprintf('\nErro Médio Absoluto entre Temperatura Real e Modelo: %.2f °C\n', erro_medio);

% Salvar gráfico
saveas(gcf, 'Grafico_TCLab_ModeloNL.png');

% Salvar dados em TXT
dados = [tm, T1, T1_ModeloNL, Q1];
header = {'Tempo_s', 'Temperatura_Real_C', 'Temperatura_Modelo_C', 'Potencia_pct'};
T = array2table(dados, 'VariableNames', header);
writetable(T, 'Dados_simulacao_ModeloNL.txt', 'Delimiter', '\t');

disp('Simulação concluída.');

%% Função de desligamento seguro
function desligarTCLab()
    try
        h1(0);
        led(0);
        disp('Dispositivos TCLab desligados com segurança.');
    catch
        disp('Erro ao tentar desligar os dispositivos.');
    end
end
